test/boolean_complex_test.cpp

```diff
@@ -1030,4 +1031,159 @@ TEST(BooleanComplex, HullMask) {
   MeshGL mesh = ret.GetMeshGL();
 }
 
+TEST(BooleanComplex, SimpleOffset) {
+  std::string file = __FILE__;
+  std::string dir = file.substr(0, file.rfind('/'));
+  MeshGL seeds = ImportMesh(dir + "/models/" + "Generic_Twin_91.1.t0.glb");
+  EXPECT_TRUE(seeds.NumTri() > 10);
+  EXPECT_TRUE(seeds.NumVert() > 10);
+  // Find unique edges
+  std::set<std::pair<int, int>> edges;
+  for (size_t i = 0; i < seeds.NumTri(); i++) {
+    int eind[3];
+    for (int j = 0; j < 3; j++) eind[j] = seeds.triVerts[i * 3 + j];
+    int e11, e12, e21, e22, e31, e32;
+    e11 = (eind[0] < eind[1]) ? eind[0] : eind[1];
+    e12 = (eind[1] < eind[0]) ? eind[0] : eind[1];
+    e21 = (eind[1] < eind[2]) ? eind[1] : eind[2];
+    e22 = (eind[2] < eind[1]) ? eind[1] : eind[2];
+    e31 = (eind[2] < eind[0]) ? eind[2] : eind[0];
+    e32 = (eind[0] < eind[2]) ? eind[2] : eind[0];
+    edges.insert(std::make_pair(e11, e12));
+    edges.insert(std::make_pair(e21, e22));
+    edges.insert(std::make_pair(e31, e32));
+  }
+  manifold::Manifold c;
+  // Vertex Spheres
+  Manifold sph = Manifold::Sphere(1, 8);
+  for (size_t i = 0; i < seeds.NumVert(); i++) {
+    vec3 vpos(seeds.vertProperties[3 * i + 0], seeds.vertProperties[3 * i + 1],
+              seeds.vertProperties[3 * i + 2]);
+    Manifold vsph = sph.Translate(vpos);
+    manifold::Manifold left = c;
+    manifold::Manifold right(vsph);
+    if (!right.NumTri()) continue;
+    c = left.Boolean(right, manifold::OpType::Add);
+    EXPECT_TRUE(c.NumTri() > 0);
+  }
+  // Edge Cylinders
+  std::set<std::pair<int, int>>::iterator e_it;
+  for (e_it = edges.begin(); e_it != edges.end(); ++e_it) {
+    vec3 ev1 = vec3(seeds.vertProperties[3 * e_it->first + 0],
+                    seeds.vertProperties[3 * e_it->first + 1],
+                    seeds.vertProperties[3 * e_it->first + 2]);
+    vec3 ev2 = vec3(seeds.vertProperties[3 * e_it->second + 0],
+                    seeds.vertProperties[3 * e_it->second + 1],
+                    seeds.vertProperties[3 * e_it->second + 2]);
+    vec3 edge = ev2 - ev1;
+    double len = la::length(edge);
+    if (len < std::numeric_limits<float>::min()) continue;
+    // TODO - workaround, shouldn't be necessary
+    if (len < 0.03) continue;
+    manifold::Manifold origin_cyl = manifold::Manifold::Cylinder(len, 1, 1, 8);
+    vec3 evec(-1 * edge.x, -1 * edge.y, edge.z);
+    manifold::Manifold rotated_cyl =
+        origin_cyl.Transform(manifold::RotateUp(evec));
+    manifold::Manifold right = rotated_cyl.Translate(ev1);
+    if (!right.NumTri()) continue;
+    manifold::Manifold left = c;
+    c = left.Boolean(right, manifold::OpType::Add);
+    EXPECT_TRUE(c.NumTri() > 0);
+  }
+  // Triangle Volumes
```

Comment:
Have you tried `Manifold::Extrude`?

Reply:
Oh,  to create the triangle based shape?  I hadn't yet, actually - I can give that a go if you think it'll be simpler.

Reply:
Up to you - let's just try to shrink the code and make it readable as much as we can.

Reply:
So looking at the same example you pointed me to for the Cylinder logic, it looks like you made a 2D simple polygon triangle, extruded it, and then used Warp and a matrix of the triangle positions to make copies of the extruded triangle at the appropriate locations?